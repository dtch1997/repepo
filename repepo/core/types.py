from typing import List, NamedTuple, Union

from transformers import PreTrainedTokenizerBase, PreTrainedModel
from repepo.repe.rep_control_reading_vec import WrappedReadingVecModel

# Placeholder type definitions
Model = Union[PreTrainedModel, WrappedReadingVecModel]
Tokenizer = PreTrainedTokenizerBase


# Base types
class Example(NamedTuple):
    instruction: str
    input: str
    output: str


class RepExample(NamedTuple):
    instruction: str 
    input: str
    output: str # for repe, we do not need an ouput, but this is not necessarily going to be the same as direction. Eg. we could have an example for 'the amount of bias is' and the output could be 'high' or 'low', not True or False
    id: int # for repe reading, we have two examples per id, one for each direction
    direction: bool # True for positive, False for negative


class Completion(NamedTuple):
    prompt: str
    response: str


Dataset = Union[List[Example], List[RepExample]]
