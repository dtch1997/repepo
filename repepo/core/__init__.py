from .types import BaseDataset
from .types import BasePipeline
