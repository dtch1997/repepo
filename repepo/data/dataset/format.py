import abc 
import random

from typing import Dict, Any, NewType, List

Example = NewType('Example', Dict[str, Any])
# Example: 
#   - a dict with keys 'instruction', 'input', 'output'
#   - could also have additional metadata

Completion = NewType('Completion', Dict[str, Any])
# Completion: 
#   - a dict with keys 'prompt', 'response'

def completion_to_str(completion: Completion) -> str:
    return completion['prompt'] + completion['response']

class AbstractFormatter(abc.ABC):
        
    def format_example(self, example: Example, **kwargs) -> Completion:
        pass

    def apply(self, examples: List[Example]) -> List[Completion]:
        completions = []
        for example in examples:
            completion = self.format_example(example)
            completions.append(completion)
        return completions
    
class QAFormatter(AbstractFormatter):
    """ Format as a simple QA pair. """
    QUESTION = "Q: {instruction} {input}"
    ANSWER = "A: {output}"

    def format_example(self, example: Dict[str, Any], **kwargs):
        del kwargs
        prompt = self.QUESTION.format_map(example)
        response = self.ANSWER.format_map(example)
        return {
            'prompt': prompt, 
            'response': response
        }


class InstructionFormatter(AbstractFormatter):
    """ Instruction formatter used for fine-tuning Alpaca. """

    PROMPT_INPUT: str = (
        "Below is an instruction that describes a task, paired with an input that provides further context. "
        "Write a response that appropriately completes the request.\n\n"
        "### Instruction:\n{instruction}\n\n### Input:\n{input}\n\n### Response:"
    )
    PROMPT_NO_INPUT: str = (
        "Below is an instruction that describes a task. "
        "Write a response that appropriately completes the request.\n\n"
        "### Instruction:\n{instruction}\n\n### Response:"
    )

    def format_example(self, example: Dict[str, Any], **kwargs):
        del kwargs
        if 'input' in example and bool(example['input']):
            prompt = self.PROMPT_INPUT.format_map(example)
        else:
            prompt = self.PROMPT_NO_INPUT.format_map(example)
        response = example['output']
        return {
            'prompt': prompt, 
            'response': response
        }

class FewShotPrompter:
    """ Compose examples few-shot """
    
    def __init__(self, n_few_shot_examples: int = 2):
        self.n_few_shot_examples = n_few_shot_examples
    
    def apply(self, completions: List[Completion]) -> List[Completion]:

        output_completions = []

        for i, completion in enumerate(completions):

            # Sample different completions for context
            few_shot_examples = []
            selected_idxes = [i, ]        
            for _ in range(self.n_few_shot_examples):
                done = False
                while not done:
                    idx = random.randint(0, len(completions) - 1) # range inclusive
                    done = idx not in selected_idxes
                few_shot_examples.append(
                    completion_to_str(completions[idx])
                )
                selected_idxes.append(idx)

            # Concatenate completions
            prompt = '\n'.join(few_shot_examples) + completion['prompt']
            response = completion['response']
            output_completions.append(dict(
                prompt=prompt, 
                response=response
            ))

        return output_completions