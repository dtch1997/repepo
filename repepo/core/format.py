import abc
from typing import List

from repepo.core.types import Completion
from repepo.core.types import Example


class AbstractFormatter(abc.ABC):
    """Describes how to format examples as completions"""

    @abc.abstractmethod
    def apply(self, example: Example, **kwargs) -> str:
        raise NotImplementedError()

    def apply_list(self, examples: List[Example]) -> List[Completion]:
        completions = []
        for example in examples:
            completion = self.apply(example)
            completions.append(completion)
        return completions


class InputOutputFormatter(AbstractFormatter):
    """Format as a simple input-output pair."""

    PROMPT_TEMPLATE = "Input: {instruction} {input} \n" "Output: "

    def apply(self, example: Example, **kwargs):
        del kwargs
        return Completion(
            prompt=self.PROMPT_TEMPLATE.format(
                instruction=example.instruction, input=example.input
            ),
            response=example.output,
        )


class InstructionFormatter(AbstractFormatter):
    """Instruction formatter used for fine-tuning Alpaca."""

    PROMPT_INPUT: str = (
        "Below is an instruction that describes a task, paired with an input that provides further context. "
        "Write a response that appropriately completes the request.\n\n"
        "### Instruction:\n{instruction}\n\n### Input:\n{input}\n\n### Response:"
    )
    PROMPT_NO_INPUT: str = (
        "Below is an instruction that describes a task. "
        "Write a response that appropriately completes the request.\n\n"
        "### Instruction:\n{instruction}\n\n### Response:"
    )

    def apply(self, example: Example, **kwargs):
        del kwargs
        if bool(example.input):
            prompt = self.PROMPT_INPUT.format(
                instruction=example.instruction, input=example.input
            )
        else:
            prompt = self.PROMPT_NO_INPUT.format_map(
                {"instruction": example.instruction}
            )
        response = example.output
        return Completion(prompt=prompt, response=response)
