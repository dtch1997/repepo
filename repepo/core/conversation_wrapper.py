from dataclasses import dataclass, field

from .types import Completion, Example
from .format import Formatter


def completion_to_str(completion: Completion) -> str:
    return completion.prompt.rstrip() + " " + completion.response.lstrip()


@dataclass
class ConversationWrapper:
    """Wraps a conversation, summarizing a list of messages into a single completion."""

    template: str = "{conversation}"
    conversation_history: list[Example] = field(default_factory=list)
    msg_separator: str = "\n"

    def wrap(self, formatter: Formatter, message: Example) -> Completion:
        completions = formatter.format_conversation(
            [*self.conversation_history, message]
        )
        prefix_completions = completions[:-1]
        final_completion = completions[-1]
        convo_prefix = self.msg_separator.join(
            [completion_to_str(c) for c in prefix_completions]
        )
        convo_str = final_completion.prompt
        if len(convo_prefix) > 0:
            convo_str = convo_prefix + self.msg_separator + final_completion.prompt
        prompt = self.template.format(conversation=convo_str)
        return Completion(prompt=prompt, response=final_completion.response)
