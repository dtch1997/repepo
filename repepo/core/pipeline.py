from dataclasses import dataclass, field
from typing import Any, Optional

from transformers import GenerationConfig

from .types import Example, Model, Tokenizer
from .prompt import AbstractPrompter, IdentityPrompter
from .format import AbstractFormatter, InputOutputFormatter


@dataclass
class Pipeline:
    """Generation pipeline"""

    model: Model
    tokenizer: Tokenizer
    prompter: AbstractPrompter = field(default_factory=IdentityPrompter)
    formatter: AbstractFormatter = field(default_factory=InputOutputFormatter)

    def build_generation_prompt(self, example: Example) -> str:
        """Build a prompt for generation"""
        completion = self.formatter.apply(example)
        completion = self.prompter.apply(completion)
        return completion.prompt

    def generate(
        self,
        example: Example,
        config: Optional[GenerationConfig] = None,
        remove_base_prompt: bool = True,
    ) -> str:
        """Generate a completion for a given example"""
        base_prompt = self.build_generation_prompt(example)
        inputs: Any = self.tokenizer(base_prompt, return_tensors="pt")
        inputs = inputs.to(self.model.device)
        outputs = self.model.generate(
            **inputs,
            config=config,
        )[0]
        outputs_str = self.tokenizer.decode(outputs, skip_special_tokens=True)
        if remove_base_prompt:
            return outputs_str.replace(base_prompt, "")
        return outputs_str
