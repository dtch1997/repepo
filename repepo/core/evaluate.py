# pyright: strict, reportMissingTypeStubs=false


from abc import ABC, abstractmethod
from contextlib import AbstractContextManager, ExitStack, contextmanager
from dataclasses import dataclass
from statistics import mean
from tqdm import tqdm
from typing import Callable, Iterable, Sequence
from repepo.core.hook import SteeringHook
from repepo.core.pipeline import TextProbs
from repepo.core.types import Example
from repepo.core.pipeline import Pipeline

import numpy as np
import logging

# eval_hooks allow us to do custom stuff to the pipeline only during evaluation
EvalHook = Callable[[Pipeline], AbstractContextManager[None]]


def print_first_example() -> EvalHook:
    """Eval hook that prints the first example"""

    @contextmanager
    def print_first_example_hook(pipeline: Pipeline):
        try:
            pipeline.print_first_example = True
            yield
        finally:
            pipeline.print_first_example = False

    return print_first_example_hook


def update_completion_template_at_eval(new_template: str) -> EvalHook:
    """Eval hook that changes the completion template for the duration of the evaluation"""

    @contextmanager
    def update_completion_template_hook(pipeline: Pipeline):
        original_template = pipeline.formatter.completion_template
        try:
            pipeline.formatter.completion_template = new_template
            yield
        finally:
            pipeline.formatter.completion_template = original_template

    return update_completion_template_hook


def set_repe_direction_multiplier_at_eval(multiplier: float) -> EvalHook:
    """Eval hook that changes the repetition penalty multiplier for the duration of the evaluation"""

    @contextmanager
    def set_repe_direction_multiplier_hook(pipeline: Pipeline):
        repe_hooks = [hook for hook in pipeline.hooks if isinstance(hook, SteeringHook)]
        if len(repe_hooks) != 1:
            raise ValueError(
                "pipeline must have exactly one SteeringHook to set repe multiplier"
            )
        repe_hook = repe_hooks[0]
        original_multiplier = repe_hook.direction_multiplier
        try:
            repe_hook.direction_multiplier = multiplier
            yield
        finally:
            repe_hook.direction_multiplier = original_multiplier

    return set_repe_direction_multiplier_hook


def select_repe_layer_at_eval(layer: int) -> EvalHook:
    """Eval hook that changes layer to steer for the duration of the evaluation"""

    @contextmanager
    def set_repe_layer_hook(pipeline: Pipeline):
        repe_hooks = [hook for hook in pipeline.hooks if isinstance(hook, SteeringHook)]
        if len(repe_hooks) != 1:
            raise ValueError(
                "pipeline must have exactly one SteeringHook to set repe multiplier"
            )
        steering_vector = repe_hooks[0].steering_vector
        if layer not in steering_vector.layer_activations:
            raise ValueError(f"layer {layer} not found in steering vector")
        original_layer_activations = steering_vector.layer_activations
        try:
            steering_vector.layer_activations = {
                layer: steering_vector.layer_activations[layer]
            }
            yield
        finally:
            steering_vector.layer_activations = original_layer_activations

    return set_repe_layer_hook


@dataclass
class EvalPrediction:
    example: Example
    positive_output_prob: TextProbs
    negative_output_prob: TextProbs


@dataclass
class EvalResult:
    predictions: list[EvalPrediction]
    metrics: dict[str, float]


class Evaluator(ABC):
    requires_generation: bool = False
    requires_probs: bool = False

    @abstractmethod
    def get_metric_name(self) -> str:
        raise NotImplementedError

    @abstractmethod
    def score_prediction(self, prediction: EvalPrediction) -> float:
        raise NotImplementedError

    def __call__(self, predictions: Sequence[EvalPrediction]) -> dict[str, float]:
        pred_results = [self.score_prediction(pred) for pred in predictions]
        return {self.get_metric_name(): mean(pred_results)}


class MultipleChoiceAccuracyEvaluator(Evaluator):
    """
    Evaluator that scores multiple choice examples by computing the probability of
    the correct output and comparing it to the probability of the incorrect outputs.
    """

    requires_probs = True

    def get_metric_name(self) -> str:
        return "mcq_acc"

    def score_prediction(self, prediction: EvalPrediction) -> float:
        """Score a single prediction, 1 if correct, 0 otherwise."""

        # the output might be longer than the expected depending on how many tokens we generate
        # so just verify that the expected output is a prefix of the generated output
        positive_output_prob = prediction.positive_output_prob.sum_logprobs
        negative_output_prob = prediction.negative_output_prob.sum_logprobs
        return 1.0 if positive_output_prob > negative_output_prob else 0.0


class LogitDifferenceEvaluator(Evaluator):
    """
    Evaluator that scores multiple choice examples by computing the average difference
    in logit between the correct and incorrect outputs.
    """

    requires_probs = True

    def get_metric_name(self) -> str:
        return "logit_diff"

    def score_prediction(self, prediction: EvalPrediction) -> float:
        """Score a single prediction based on difference in sum of logits."""

        # calculate difference in logits
        positive_output_logit = prediction.positive_output_prob.sum_logits
        negative_output_logit = prediction.negative_output_prob.sum_logits
        return positive_output_logit - negative_output_logit


class NormalizedPositiveProbabilityEvaluator(Evaluator):
    """
    Evaluator that scores multiple choice examples by computing the
    normalized probability of the positive output
    """

    requires_probs = True

    def get_metric_name(self) -> str:
        return "pos_prob"

    def score_prediction(self, prediction: EvalPrediction) -> float:
        """
        Normalize the probabilities of positive and negative outputs relative to each other
        NOTE: This returns actual probabilities, not logprobs
        """

        # calculate normalized logprobs
        positive_output_logprob = prediction.positive_output_prob.sum_logprobs
        negative_output_logprob = prediction.negative_output_prob.sum_logprobs

        # normalize by max to avoid underflow?
        max_logprob = max(positive_output_logprob, negative_output_logprob)
        positive_output_logprob = positive_output_logprob - max_logprob
        negative_output_logprob = negative_output_logprob - max_logprob

        # Calculate normalized probability
        positive_output_prob = np.exp(positive_output_logprob)
        negative_output_prob = np.exp(negative_output_logprob)
        return positive_output_prob / (positive_output_prob + negative_output_prob)


def evaluate(
    pipeline: Pipeline,
    dataset: Iterable[Example],
    evaluators: Sequence[Evaluator],
    # these eval_hooks allow us to do custom stuff to the pipeline only during evaluation,
    # e.g. mess with the formatter to use CAA's special answer format
    eval_hooks: Sequence[EvalHook] = [],
    show_progress: bool = True,
    tqdm_desc: str = "Evaluating",
    logger: logging.Logger | None = None,
) -> EvalResult:
    # evaluate
    predictions: list[EvalPrediction] = []

    with ExitStack() as stack:
        for eval_hook in eval_hooks:
            stack.enter_context(eval_hook(pipeline))
        # TODO: support batching
        for i, example in enumerate(
            tqdm(dataset, disable=not show_progress, desc=tqdm_desc)
        ):
            if logger is not None and i == 0:
                logger.info(
                    f"Example full prompt: \n {pipeline.build_full_prompt(example.positive)}"
                )
            positive_probs = pipeline.calculate_output_logprobs(example.positive)
            negative_probs = pipeline.calculate_output_logprobs(example.negative)

            predictions.append(
                EvalPrediction(
                    example=example,
                    positive_output_prob=positive_probs,
                    negative_output_prob=negative_probs,
                )
            )
        metrics: dict[str, float] = {}
        for evaluator in evaluators:
            metrics.update(evaluator(predictions))
        return EvalResult(predictions, metrics)
    raise RuntimeError("Should never get here")
