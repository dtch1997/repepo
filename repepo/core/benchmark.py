# pyright: strict, reportMissingTypeStubs=false

from dataclasses import dataclass
from typing import Optional, Sequence

from transformers.generation import GenerationConfig

from repepo.algorithms.base import Algorithm
from repepo.core.evaluate import (
    EvalHook,
    EvalResult,
    Evaluator,
    evaluate,
)
from repepo.core.format import Formatter, InputOutputFormatter
from repepo.core.pipeline import Pipeline

from repepo.core.types import Dataset, Model, Tokenizer


@dataclass
class Benchmark:
    name: str
    train_dataset: Dataset
    test_dataset: Dataset
    evaluators: list[Evaluator]


def train_benchmark(
    model: Model,
    tokenizer: Tokenizer,
    algorithms: Sequence[Algorithm],
    benchmark: Benchmark,
    formatter: Optional[Formatter] = None,
) -> Pipeline:
    pipeline = Pipeline(
        model,
        tokenizer,
        formatter=formatter or InputOutputFormatter(),
    )
    for algorithm in algorithms:
        # Re-initialize pipeline, which gets destructively modified
        # TODO: do something with outputs?
        _ = algorithm.run(pipeline, benchmark.train_dataset)
    return pipeline


def evaluate_benchmark(
    pipeline: Pipeline,
    benchmark: Benchmark,
    generation_config: Optional[GenerationConfig] = None,
    # these eval_hooks allow us to do custom stuff to the pipeline only during evaluation,
    # e.g. mess with the formatter to use CAA's special answer format
    eval_hooks: list[EvalHook] = [],
    show_progress: bool = True,
    tqdm_desc: str = "Evaluating",
) -> EvalResult:
    # evaluate
    return evaluate(
        pipeline,
        dataset=benchmark.test_dataset,
        evaluators=benchmark.evaluators,
        generation_config=generation_config,
        eval_hooks=eval_hooks,
        show_progress=show_progress,
        tqdm_desc=tqdm_desc,
    )


def train_and_evaluate_benchmark(
    model: Model,
    tokenizer: Tokenizer,
    algorithms: Sequence[Algorithm],
    benchmark: Benchmark,
    formatter: Optional[Formatter] = None,
    generation_config: Optional[GenerationConfig] = None,
    # these eval_hooks allow us to do custom stuff to the pipeline only during evaluation,
    # e.g. mess with the formatter to use CAA's special answer format
    eval_hooks: list[EvalHook] = [],
    show_progress: bool = True,
) -> EvalResult:
    # train
    pipeline = train_benchmark(
        model,
        tokenizer,
        algorithms,
        benchmark,
        formatter=formatter,
    )

    # evaluate
    return evaluate_benchmark(
        pipeline,
        benchmark,
        generation_config=generation_config,
        eval_hooks=eval_hooks,
        show_progress=show_progress,
    )
