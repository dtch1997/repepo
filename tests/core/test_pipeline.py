import torch
import scipy
import numpy as np
from textwrap import dedent
from transformers import GPTNeoXForCausalLM

from repepo.core.pipeline import Pipeline, compute_moments, compute_quantiles
from repepo.core.format import IdentityFormatter, LlamaChatFormatter
from repepo.core.types import Completion, Tokenizer
from syrupy.assertion import SnapshotAssertion


def _compute_moments_scipy(x: np.ndarray, axis: int) -> np.ndarray:
    mean = np.mean(x, axis=axis)
    std = scipy.stats.tstd(x, axis=axis, ddof=1)
    skew = scipy.stats.skew(x, axis=axis)
    kurtosis = scipy.stats.kurtosis(x, axis=axis, fisher=False)
    return np.stack([mean, std, skew, kurtosis], axis=1)


def test_compute_moments_basic():
    tensor = torch.tensor(
        [[1.0, 2.0, 3.0, 4.0, 5.0, 6.0], [2.0, 2.0, 2.0, 2.0, 2.0, 2.0]]
    )
    output = compute_moments(tensor, dim=1)
    expected_output = torch.from_numpy(_compute_moments_scipy(tensor.numpy(), axis=1))
    # NOTE: torch kurtosis does not agree with scipy kurtosis for some reason...
    # Omitted from testing for now
    torch.testing.assert_allclose(output[:, :3], expected_output[:, :3])


def test_compute_moments_edge_cases():
    # Test with a single-value tensor
    tensor = torch.tensor([[1.0]])
    expected_output = torch.tensor([[1.0, np.nan, np.nan, np.nan]])
    output = compute_moments(tensor, dim=1)
    torch.testing.assert_allclose(output[:, :3], expected_output[:, :3])

    # Test with a tensor with uniform values
    tensor = torch.full((1, 4), 3.0)
    expected_output = torch.tensor([[3.0, 0.0, np.nan, np.nan]])
    output = compute_moments(tensor, dim=1)
    torch.testing.assert_allclose(output[:, :3], expected_output[:, :3])


def test_compute_quantiles_basic():
    tensor = torch.tensor([[1.0, 2.0, 3.0, 4.0], [4.0, 3.0, 2.0, 1.0]])
    expected_output = torch.tensor(
        [[1.0, 1.75, 2.5, 3.25, 4.0], [1.0, 1.75, 2.5, 3.25, 4.0]]
    )
    output = compute_quantiles(tensor, dim=1)
    torch.testing.assert_allclose(output, expected_output)


def test_compute_quantiles_edge_cases():
    # Test with a single-value tensor
    tensor = torch.tensor([[2.0]])
    expected_output = torch.tensor([[2.0, 2.0, 2.0, 2.0, 2.0]])
    output = compute_quantiles(tensor, dim=1)
    torch.testing.assert_allclose(output, expected_output)

    # Test with non-unique values
    tensor = torch.tensor([[2.0, 2.0, 2.0, 2.0]])
    expected_output = torch.tensor([[2.0, 2.0, 2.0, 2.0, 2.0]])
    output = compute_quantiles(tensor, dim=1)
    torch.testing.assert_allclose(output, expected_output)


def test_basic_Pipeline_generate(
    model: GPTNeoXForCausalLM, tokenizer: Tokenizer
) -> None:
    pipeline = Pipeline(model, tokenizer)
    res = pipeline.generate(
        Completion(prompt="Respond A B C D", response="E"),
        generation_config=None,
    )
    # pythia-70m generates nonsense, so just verify we get something
    assert isinstance(res, str)
    assert len(res) > 0

def test_basic_Pipeline_build_generation_prompt(
    model: GPTNeoXForCausalLM, tokenizer: Tokenizer
) -> None:
    pipeline = Pipeline(model, tokenizer)
    res = pipeline.build_generation_prompt(
        Completion(prompt="Respond A B C D", response="E")
    )
    assert res == "Respond A B C D"


def test_basic_Pipeline_build_full_prompt(
    model: GPTNeoXForCausalLM, tokenizer: Tokenizer
) -> None:
    pipeline = Pipeline(model, tokenizer)
    res = pipeline.build_full_prompt(Completion(prompt="Respond A B C D", response="E"))
    assert res == "Respond A B C D E"


def test_basic_Pipeline_build_generation_prompt_with_custom_completion_template(
    model: GPTNeoXForCausalLM, tokenizer: Tokenizer
) -> None:
    formatter = IdentityFormatter(
        completion_template="Input: {prompt}\n Output: My answer is: {response}",
    )
    pipeline = Pipeline(model, tokenizer, formatter=formatter)
    res = pipeline.build_generation_prompt(
        Completion(prompt="Respond A B C D", response="E")
    )
    assert res == "Input: Respond A B C D\n Output: My answer is:"


def test_basic_Pipeline_build_full_prompt_with_custom_completion_template(
    model: GPTNeoXForCausalLM, tokenizer: Tokenizer
) -> None:
    formatter = IdentityFormatter(
        completion_template="Input: {prompt}\n Output: My answer is: {response}",
    )
    pipeline = Pipeline(model, tokenizer, formatter=formatter)
    res = pipeline.build_full_prompt(
        Completion(prompt="Respond A B C D", response="E"),
    )
    assert res == "Input: Respond A B C D\n Output: My answer is: E"


def test_icl_Pipeline_build_generation_prompt(
    model: GPTNeoXForCausalLM, tokenizer: Tokenizer, snapshot: SnapshotAssertion
) -> None:
    dataset = [
        Completion(prompt="Paris is in", response="France"),
        Completion(prompt="London is in", response="England"),
        Completion(prompt="Berlin is in", response="Germany"),
    ]
    pipeline = Pipeline(model, tokenizer)
    pipeline.conversation_history = dataset
    res = pipeline.build_generation_prompt(
        Completion(prompt="Beijing is in", response="China")
    )
    assert res == snapshot


def test_basic_pipeline_calculate_output_logprobs(
    model: GPTNeoXForCausalLM, tokenizer: Tokenizer
) -> None:
    pipeline = Pipeline(model, tokenizer)
    res = pipeline.calculate_output_logprobs(
        Completion(
            prompt="Input: Select the best answer.\nA B C D", response="Output: D"
        )
    )
    assert res.sum_logprobs < 0
    assert res.text == "Input: Select the best answer.\nA B C D Output: D"
    assert (
        "".join([tok.text for tok in res.token_probs])
        # "Input" is the first token, so the model doesn't predict this
        == ": Select the best answer.\nA B C D Output: D"
    )
    for tok in res.token_probs:
        assert tok.logprob < 0


def test_pipeline_can_reproduce_caa_train_prompts(
    model: GPTNeoXForCausalLM, tokenizer: Tokenizer
) -> None:
    # copied from a real prompt in the CAA training set, abbreviated for brevity
    target = dedent(
        """
        [INST] <<SYS>>
        You are a helpful, honest and concise assistant.
        <</SYS>>
        
        Hello, my name is Linda Johnson ...
        Choices:
        (A) Agree
        (B) Disagree [/INST] (B)
        """
    )
    formatter = LlamaChatFormatter()
    pipeline = Pipeline(model, tokenizer, formatter=formatter)
    completion = Completion(
        prompt=dedent(
            """
            Hello, my name is Linda Johnson ...
            Choices:
            (A) Agree
            (B) Disagree
            """
        ),
        response="(B)",
    )
    assert pipeline.build_full_prompt(completion).strip() == target.strip()
