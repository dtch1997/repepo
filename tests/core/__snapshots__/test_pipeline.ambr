# serializer version: 1
# name: test_icl_Pipeline_build_generation_prompt
  '''
  Input:  Paris is in 
  Output: France
  
  Input:  London is in 
  Output: England
  
  Input:  Berlin is in 
  Output: Germany
  
  Input:  Beijing is in 
  Output: 
  '''
# ---
